content/general/concepts/programming-optimization/programming-optimization.md

```diff
@@ -0,0 +1,213 @@
+---
+Title: 'Programming Optimization'
+Description: 'Optimization in programming is the process of modifying a software system to execute faster and become more efficient by minimizing the use of available computing resources.'
+Subjects:
+  - 'Web Development'
+Tags:
+  - 'Optimization in Programing'
+  - 'Programming Optimization'
+CatalogContent:
+  - 'paths/computer-science'
+  - 'web-development'
+---
+
+**Programming Optimization** is the process of modifying a software system to amplify its functionality, making the application execute faster and more efficiently by minimizing the use of computing resources. In other words, a computer program may be optimized to reduce execution time, memory usage, disk space, bandwidth, and even draw less power. This usually requires a trade-off, where one aspect is implemented at the expense of others, e.g., increasing the size of cache improves run-time performance but also increases memory consumption.
+
+It has become an important focal point since today’s users appreciate a highly performant experience and have significantly higher expectations. It has become a deciding factor of the success of modern business in today’s digital market. Optimization in programming is a key element in the improvement of such performance.
+
+## Programming Optimization
+
+Optimization in programming is an extensive topic, it relies on many different metrics, focuses and areas within the application depending on specific situations.
+
+Although consideration and awareness of performance issues of design and implementation should be factored into every part of the development cycle, it should not be the primary concern within the first iteration. These can potentially complicate the design and/or implementation, hence should be left until the final iteration of the application. 
+
+Techniques used can be broken up among various scopes which can affect anything from a single statement to the entire program. Generally speaking, locally scoped techniques are easier to implement than global ones, though they may result in smaller gains. 
+
+These are important points to target for optimization in programming:
+
+- Reducing code complexity
+- Avoiding unnecessary for-loops
+- Using data structures similar to like sorting, string concatenation
+- Avoiding unnecessary functions
+- Avoiding dots
+- Initializing dictionary elements
+- Import statement overhead
+- Data aggregation
+- Re-map functions at runtime
+- Profiling code
+
+## Levels of Optimization
+
+Every application, generally, is composed of 2 areas where programming optimizations can be applied within levels of development:
+
+**Client Side**, which is the side seen on the web browser by the user, includes: 
+- JavaScript that runs in the browser 
```

Comment:
To maintain consistency, capitalize the first letter of all the bullet points in this entry.